tests/test_analysis/test_outliers/test_iqr_outliers.py

```diff
@@ -0,0 +1,236 @@
+)
+def test_sliding_window_params(window_size, stride, error):
+    with pytest.raises(ValueError, match=error):
+        _ = _sliding_window(x=np.empty, window_size=window_size, stride=stride)
+
+
+@pytest.mark.parametrize("data", (np.arange(9),))
+@pytest.mark.parametrize(
+    "window_size, stride, answer",
+    (
+        (3, 1, np.array([[0, 1, 2], [1, 2, 3], [2, 3, 4], [3, 4, 5], [4, 5, 6], [5, 6, 7], [6, 7, 8]])),
+        (2, 1, np.array([[0, 1], [1, 2], [2, 3], [3, 4], [4, 5], [5, 6], [6, 7], [7, 8]])),
+        (4, 1, np.array([[0, 1, 2, 3], [1, 2, 3, 4], [2, 3, 4, 5], [3, 4, 5, 6], [4, 5, 6, 7], [5, 6, 7, 8]])),
+        (3, 3, np.array([[0, 1, 2], [3, 4, 5], [6, 7, 8]])),
+        (7, 2, np.array([[0, 1, 2, 3, 4, 5, 6], [2, 3, 4, 5, 6, 7, 8]])),
+        (1, 1, np.array([[0], [1], [2], [3], [4], [5], [6], [7], [8]])),
+        (1, 9, np.array([[8]])),
+        (9, 1, np.array([[0, 1, 2, 3, 4, 5, 6, 7, 8]])),
+    ),
+)
+def test_sliding_window(data, window_size, stride, answer):
+    res = _sliding_window(x=data, window_size=window_size, stride=stride)
+    np.testing.assert_allclose(res, answer)
+
+
+@pytest.mark.parametrize("data", (np.arange(9),))
+@pytest.mark.parametrize(
+    "window_size, stride, answer",
+    (
+        (3, 1, np.array([3, 6, 9, 12, 15, 18, 21])),
+        (2, 1, np.array([1, 3, 5, 7, 9, 11, 13, 15])),
+        (4, 1, np.array([6, 10, 14, 18, 22, 26])),
+        (3, 3, np.array([3, 12, 21])),
+        (7, 2, np.array([21, 35])),
+        (1, 1, np.array([0, 1, 2, 3, 4, 5, 6, 7, 8])),
+        (1, 9, np.array([8])),
+        (9, 1, np.array([36])),
+    ),
+)
+def test_sliding_window_decorator(data, window_size, stride, answer):
+    apply_func = sliding_window_decorator(lambda x, idxs: np.sum(x[idxs]))
+    res = apply_func(data, window_size=window_size, stride=stride, return_indices=False)
+    np.testing.assert_allclose(res, answer)
+
+
+def test_stl_decompose_error():
+    with pytest.raises(ValueError, match="At least one component must be set!"):
+        _ = _stl_decompose(series=pd.Series([0.0]), trend=False, seasonality=False)
+
+
+def test_const_ts(const_ts_anomal):
+    anomal = get_anomalies_iqr(const_ts_anomal)
+    assert len(anomal) == 0
+
+
+def test_iqr_outliers_missing_values(outliers_tsds):
+    with pytest.raises(ValueError, match=".* contains missing values!"):
+        _ = get_anomalies_iqr(ts=outliers_tsds)
+
+
+def test_iqr_outliers_invalid_scale(outliers_df_with_two_columns):
+    with pytest.raises(ValueError, match="Scaling parameter must be positive!"):
+        _ = get_anomalies_iqr(ts=outliers_df_with_two_columns, iqr_scale=-1.0)
+
+
+@pytest.mark.parametrize(
+    "ts_name, error",
+    (
+        (
+            "outliers_df_with_two_columns_int_timestamp",
+            "Series must have inferable frequency to autodetect period for STL!",
+        ),
+        ("outliers_df_with_two_columns_minute_freq", "freq T not understood. Please report"),
```

Comment:
Why is this some unknown freq?

Reply:
`statsmodels` implementation of `STL` uses [this](https://github.com/statsmodels/statsmodels/blob/52edda81159e557dd845cc97a8a9324915c69792/statsmodels/tsa/tsatools.py#L784) function to infer period length from frequency. Unfortunately, it doesn't support all frequencies. In such cases, the period length must be provided directly.